plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java

```diff
@@ -226,60 +241,154 @@ private RowId getRowId(int position)
 
     private Set<RowId> getDeletedRows()
     {
-        if (deletedRows != null) {
-            return deletedRows;
+        checkState(state == State.LOADED, "expected LOADED state but was %s", state);
+        verify(deletedRows != null, "deleted rows null despite LOADED state");
+        return deletedRows;
+    }
+
+    public boolean loadOrYield()
+    {
+        checkState(state != State.CLOSED, "already closed");
+
+        if (state == State.NOT_LOADED) {
+            loader = new Loader();
+            state = State.LOADING;
+        }
+
+        if (state == State.LOADING) {
+            verify(loader != null, "loader not set despite LOADING state");
+            Optional<Set<RowId>> loadedRowIds = loader.loadOrYield();
+            if (loadedRowIds.isPresent()) {
+                deletedRows = loadedRowIds.get();
+                try {
+                    loader.close();
+                }
+                catch (IOException e) {
+                    throw new TrinoException(HIVE_CURSOR_ERROR, "Failed to close deletedRows ", e);
+                }
+                loader = null;
+                state = State.LOADED;
+            }
+        }
+
+        if (state == State.LOADED) {
+            return true;
+        }
+        return false;
+    }
+
+    public void close()
+            throws IOException
+    {
+        if (state == State.CLOSED) {
+            return;
```

Comment:
technically, this is redundant. ok to keep, ok to remove

Reply:
WIll keep. A bit clener to me.